pkg/infra/spot.go

```diff
@@ -173,3 +233,244 @@ func setSpotOnGCPProviderSpec(params framework.MachineSetParams) error {
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: getMetadataMockLabels(),
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:    "metadata-mock",
+							Image:   "golang:1.14",
+							Command: []string{"/usr/local/go/bin/go"},
+							Args: []string{
+								"run",
+								"/mock/metadata_mock.go",
+								fmt.Sprintf("--provider=%s", platform),
+								fmt.Sprintf("--listen-addr=0.0.0.0:%d", metadataServiceMockPort),
+							},
+							Env: []corev1.EnvVar{
+								{
+									Name:  "GOCACHE",
+									Value: "/go/.cache",
+								},
+							},
+							VolumeMounts: []corev1.VolumeMount{
+								{
+									Name:      "mock-server",
+									MountPath: "/mock",
+								},
+							},
+						},
+					},
+					DNSPolicy: corev1.DNSClusterFirst,
+					Volumes: []corev1.Volume{
+						{
+							Name: "mock-server",
+							VolumeSource: corev1.VolumeSource{
+								ConfigMap: &corev1.ConfigMapVolumeSource{
+									LocalObjectReference: corev1.LocalObjectReference{
+										Name: metadataServiceMockName + "-config",
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func getMetadataMockService() *corev1.Service {
+	return &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      metadataServiceMockName + "-service",
+			Namespace: framework.MachineAPINamespace,
+			Labels:    getMetadataMockLabels(),
+		},
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{
+				{
+					Name:       "http",
+					Port:       metadataServiceMockPort,
+					Protocol:   "TCP",
+					TargetPort: intstr.FromInt(metadataServiceMockPort),
+				},
+			},
+			Selector:        getMetadataMockLabels(),
+			SessionAffinity: corev1.ServiceAffinityNone,
+			ClusterIP:       "None",
+			Type:            corev1.ServiceTypeClusterIP,
+		},
+	}
+}
+
+func getMetadataMockConfigMap() (*corev1.ConfigMap, error) {
+	// Load relative to the test execution directory
+	data, err := ioutil.ReadFile("./infra/mock/metadata_mock.go")
+	if err != nil {
+		return nil, err
+	}
+
+	return &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      metadataServiceMockName + "-config",
+			Namespace: framework.MachineAPINamespace,
+			Labels:    getMetadataMockLabels(),
+		},
+		BinaryData: map[string][]byte{
+			"metadata_mock.go": data,
+		},
+	}, nil
+}
+
+const (
+	terminationSimulatorName = "termination-simulator"
+)
+
+func getTerminationSimulatorPod(nodeName string) *corev1.Pod {
+	script := fmt.Sprintf(`apk update && apk add iptables bind-tools;
+export SERVICE_IP=$(dig +short %s.${NAMESPACE}.svc.cluster.local);
+if [ -z ${SERVICE_IP} ]; then echo "No service IP"; exit 1; fi;
+iptables-nft -t nat -A OUTPUT -p tcp -d 169.254.169.254 -j DNAT --to-destination ${SERVICE_IP}:%d;
+iptables-nft -t nat -A POSTROUTING -j MASQUERADE;
+echo "Redirected metadata service to ${SERVICE_IP}:%d";`, metadataServiceMockName+"-service", metadataServiceMockPort, metadataServiceMockPort)
```

Comment:
this part looks a little crazy here, but i don't have a suggestion to make it cleaner.

Reply:
iptables stuff can be clean? :D

Reply:
I could switch the fmt string to just being a static string and use environment variables for the substitutions, that might make it a little cleaner

Reply:
i don't know, it seems like we are just pushing things around if we move. i'm fine to leave this, it's not a blocker for me just something i noticed.

Reply:
I moved it in a separate commit, so can undo easily, wdyt, this way or the old way? Not sure how else to make it cleaner

Reply:
i think this is an improvement for readability. the env variables do make it easier for me to understand what is happening. big +1 here, thanks!